cobalt/media/base/metrics_provider.cc

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/media/base/metrics_provider.h"
+
+#include "base/logging.h"
+#include "base/metrics/histogram_functions.h"
+
+
+namespace cobalt {
+namespace media {
+
+using starboard::ScopedLock;
+
+
+MediaMetricsProvider::~MediaMetricsProvider() {
+  if (!IsInitialized()) return;
+  ReportPipelineUMA();
+}
+
+std::string MediaMetricsProvider::GetUMANameForAVStream(
+    const PipelineInfo& player_info) {
+  constexpr char kPipelineUmaPrefix[] = "Media.PipelineStatus.AudioVideo.";
+  std::string uma_name = kPipelineUmaPrefix;
+  if (player_info.video_codec == VideoCodec::kVP8)
+    uma_name += "VP8.";
+  else if (player_info.video_codec == VideoCodec::kVP9)
+    uma_name += "VP9.";
+  else if (player_info.video_codec == VideoCodec::kH264)
+    uma_name += "H264.";
+  else if (player_info.video_codec == VideoCodec::kAV1)
+    uma_name += "AV1.";
+  else
+    return uma_name + "Other";
+
+  if (player_info.video_pipeline_info.has_decrypting_demuxer_stream)
+    uma_name += "DDS.";
+
+  // Note that HW essentially means 'platform' anyway. MediaCodec has been
+  // reported as HW forever, regardless of the underlying platform
+  // implementation.
+  uma_name += player_info.video_pipeline_info.is_platform_decoder ? "HW" : "SW";
+
+  return uma_name;
+}
+
+void MediaMetricsProvider::ReportPipelineUMA() {
+  ScopedLock scoped_lock(mutex_);
+  if (uma_info_.has_video && uma_info_.has_audio) {
+    base::UmaHistogramExactLinear(GetUMANameForAVStream(uma_info_),
+                                  uma_info_.last_pipeline_status,
+                                  ::media::PIPELINE_STATUS_MAX + 1);
+  } else if (uma_info_.has_audio) {
+    base::UmaHistogramExactLinear("Media.PipelineStatus.AudioOnly",
+                                  uma_info_.last_pipeline_status,
+                                  ::media::PIPELINE_STATUS_MAX + 1);
+  } else if (uma_info_.has_video) {
+    base::UmaHistogramExactLinear("Media.PipelineStatus.VideoOnly",
+                                  uma_info_.last_pipeline_status,
+                                  ::media::PIPELINE_STATUS_MAX + 1);
+  } else {
+    // Note: This metric can be recorded as a result of normal operation with
+    // Media Source Extensions. If a site creates a MediaSource object but never
+    // creates a source buffer or appends data, PIPELINE_OK will be recorded.
+    base::UmaHistogramExactLinear("Media.PipelineStatus.Unsupported",
+                                  uma_info_.last_pipeline_status,
+                                  ::media::PIPELINE_STATUS_MAX + 1);
+  }
+
+  // Report whether video decoder fallback happened for each video codec, but
+  // only if a video decoder was reported.
+  if (uma_info_.video_pipeline_info.decoder_type !=
+      VideoDecoderType::kUnknown) {
+    base::UmaHistogramBoolean("Media.VideoDecoderFallback." +
```

Comment:
I think histograms.xml is missing both the VideoDecoderFallback and HasEveryPlayed metadata

Reply:
removed them. no longer needed